### PR TITLE
Improve GeoDjango type-hints

### PR DESCRIPTION
**Pull request by ciscorn**
# I have made things!

- Provide more specific type-hints for GeoDjango modules.
- Fixed some missing exports in `contrib/gis/**/__init__.py`
- Removed unintentional re-exports in many modules.

## Related issues

No

### CODE REVIEW

**Review by sobolevn — COMMENTED** [review]
Oh wow! Thanks a lot! This is amazing!

But, I literally have 0 experience with `gis`.
Maybe @intgr or @adamchainz have more knowledge?

**Comment by intgr** [comment]
Nope, I have no experience with Django GIS.

**Review by intgr — APPROVED** [review]
Looks good on a cursory read. As noted above, the TypedDjango team doesn't use GeoDjango. I'm inclined to just merge this as is, it's filling out types that used to be just `Any`.

Changes like this are rarely perfect and we shouldn't expect them to be. If it causes regressions for users, they will be reported and can be improved.

**Comment by intgr** [comment]
There was some overlap between this PR and #1265 (which just got merged to master).

I have pushed a merge with the latest master.

File: django-stubs/contrib/gis/gdal/datasource.pyi [hunk 1]
```diff
@@ -1,16 +1,18 @@
 from typing import Any
 
 from django.contrib.gis.gdal.base import GDALBase as GDALBase
+from django.contrib.gis.gdal.driver import Driver as Driver
+from django.contrib.gis.gdal.layer import Layer as Layer
```
Comment (intgr) [inline django-stubs/contrib/gis/gdal/datasource.pyi]: These probably shouldn't be re-exported. The `GDALBase` above (which you didn't touch, to be fair) looks wrong too.

```suggestion
from django.contrib.gis.gdal.base import GDALBase
from django.contrib.gis.gdal.driver import Driver
from django.contrib.gis.gdal.layer import Layer
```

File: django-stubs/contrib/gis/gdal/feature.pyi [hunk 2]
```diff
@@ -1,27 +1,30 @@
 from typing import Any
 
 from django.contrib.gis.gdal.base import GDALBase as GDALBase
+from django.contrib.gis.gdal.field import Field as Field
+from django.contrib.gis.gdal.geometries import OGRGeometry as OGRGeometry
+from django.contrib.gis.gdal.geomtype import OGRGeomType as OGRGeomType
```
Comment (intgr) [inline django-stubs/contrib/gis/gdal/feature.pyi]: Looks like ordinary imports, shouldn't be re-exported.

```suggestion
from django.contrib.gis.gdal.base import GDALBase
from django.contrib.gis.gdal.field import Field
from django.contrib.gis.gdal.geometries import OGRGeometry
from django.contrib.gis.gdal.geomtype import OGRGeomType
```

File: django-stubs/contrib/gis/gdal/geometries.pyi [hunk 3]
```diff
@@ -1,136 +1,142 @@
-from typing import Any
+from typing import Any, AnyStr
 
 from django.contrib.gis.gdal import CoordTransform, SpatialReference
 from django.contrib.gis.gdal.base import GDALBase as GDALBase
+from django.contrib.gis.gdal.envelope import Envelope as Envelope
+from django.contrib.gis.gdal.geomtype import OGRGeomType as OGRGeomType
```
Comment (intgr) [inline django-stubs/contrib/gis/gdal/geometries.pyi]: Looks like ordinary imports, shouldn't be re-exported.

```suggestion
from django.contrib.gis.gdal.base import GDALBase
from django.contrib.gis.gdal.envelope import Envelope
from django.contrib.gis.gdal.geomtype import OGRGeomType
```

**Review by intgr — CHANGES_REQUESTED** [review]
After looking into this some more, there are lots of re-exports that look like they shouldn't be re-exports.

Usually Django only has re-exports in `__init__.py` files. Also items in explicit module `__all__ = [...]` should be re-exported. The remaining typically should not be.

**Comment by intgr** [comment]
PS: I don't expect you to fix all the pre-existing incorrect re-exports. Just fix the new ones you have added in this PR.

File: django-stubs/contrib/gis/utils/layermapping.pyi [hunk 4]
```diff
@@ -2,8 +2,8 @@ from collections.abc import Mapping
 from pathlib import Path
 from typing import Any, Protocol
 
+import django.contrib.gis.gdal.field
 from django.contrib.gis.gdal import DataSource, OGRGeomType
-from django.contrib.gis.gdal.field import Field as OGRField
```
Comment (intgr) [inline django-stubs/contrib/gis/utils/layermapping.pyi]: `from django.contrib.gis.gdal.field import Field as OGRField` is preferrable here.

It doesn't count as a re-export; re-exports are only those that have the same name (e.g. `import Field as Field`)

Comment (intgr) [reply django-stubs/contrib/gis/utils/layermapping.pyi]: Documented here: https://typing.readthedocs.io/en/latest/source/stubs.html#imports

**Review by intgr — APPROVED** [review]
Thanks! Waiting for the CI to finish.

**Comment by intgr** [comment]
Maybe @mjakob can try this out and share some feedback.

File: django-stubs/contrib/gis/utils/layermapping.pyi (hunk 4 above)
Comment (ciscorn) [reply django-stubs/contrib/gis/utils/layermapping.pyi]: Thank you for your review!

**Pull request merged**